"New file" command

mandx opened the pull request:
This commands allows file creation anywhere in the filesystem and it will attempt to create all the subdirectories needed to realize the path.

The command will attempt to check if the file exists already and exit early if that's the case; otherwise no other checks are performed: if the operation fails for any reason, a `status_message` is emitted with the contents of the error/exception.

Not quite what's in #40, but IMO it's a good starting point. This is how it looks like in my environment:
![image](https://user-images.githubusercontent.com/535282/179842041-4407663b-997e-4c04-98f4-8235763f6863.png)
When entering a filename, the full path is editable, but initially, the filename (basename+extension) is pre-selected so one can just start typing to create an empty file (and possibly new directories) in the selected directory.
![image](https://user-images.githubusercontent.com/535282/179842229-eb53d742-c1ce-4e61-bb2c-62f07255b98f.png)

mandx committed:
"New file" command

This commands allows file creation anywhere in the filesystem and it
will attempt to create all the subtrees needed to realize the path.

The command will attempt to check if the file exists already and exit
early if that's the case; otherwise no other checks are performed: if
the operation fails for any reason, a `status_message` is emitted with
the contents of the error/exception.

braver commented:
Nifty. Normally I would defer to either existing functionality (but this clearly is more powerful) or existing packages (but AdvancedNewFile is a bit bloated, old and seemingly unmaintained... not a promising situation). 

> The command will attempt to check if the file exists already and exit early if that's the case;

I haven't tested yet but would assume that we open the new file as soon as it is created. If the file already exists we could flash a message ("Opening existing file") and open the file.

mandx commented:
> I haven't tested yet but would assume that we open the new file as soon as it is created.

Yes, that's what it does

> If the file already exists we could flash a message ("Opening existing file") and open the file.

It doesn't do that right now, but it's actually a very easy change!

mandx committed:
Attempt to open the file regardless of if it was created or not; Inform the file already exists using an alert box

mandx commented:
The change so far:

![image](https://user-images.githubusercontent.com/535282/180063715-e755a2d0-ab83-46a3-b728-2eeec8fc15d7.png)

![image](https://user-images.githubusercontent.com/535282/180064202-db6da829-8163-4e06-9f08-6991b5463bb4.png)

mandx committed:
Rewording

braver commented on Commands.sublime-commands:
```diff
@@ -20,15 +20,20 @@
 	},
 	{
-		"caption": "File: Move",
+		"caption": "File: Move…",
```
👍🏻 better!

braver commented:
LGTM. I'll give it a final test run and then merge 🚢

braver commented:
Idea:

If you right-click a file in the sidebar, or have file opened and run from the command palette, we could seed the dialog with the selected file instead of "New file". Especially when you right-click a file it feels like that would make sense.

braver commented on SideBar.py:
```diff
@@ -274,3 +274,49 @@ def on_post_window_command(self, window, command_name, args):
+
+    def on_done(self, filename):
+        if filename.endswith(os.path.sep):
+            self.window.status_message(
```
We could also interpret a new "file" ending with `/` to mean "create directory with this name". Especially since the command is "New..." and not explicitly "New file...".

braver commented:
> Idea:
> 
> If you right-click a file in the sidebar, or have file opened and run from the command palette, we could seed the dialog with the selected file instead of "New file". Especially when you right-click a file it feels like that would make sense.

Huh, I suggested as much in #40:

> If the active view has a file name, suggest the new file to be have the same name. Just like the "Move" command we could pre-select the filename so it's easy to change. This is a good starting point for creating e.g. another file of the same type, and it's super easy to create the file in any of the parent directories.

mandx committed:
The New command now supports creating directories, and the filename suggestion will use (when available) the file from the sidebar or the path of the current view

mandx commented:
> If the active view has a file name, suggest the new file to be have the same name. Just like the "Move" command we could pre-select the filename so it's easy to change. This is a good starting point for creating e.g. another file of the same type, and it's super easy to create the file in any of the parent directories.

Done! I'm also detecting when "New..." was used on a directory, so in this case, the input panel will have the generic `New file.txt` path, now the difference is in the selection, when the command is used under a directory, the whole `New file.txt` text is selected (including the extension), but if it's used on an existing file, that file's name is used, but the selection will contain only the basename, the extension will be not be pre-selected.

braver commented on Commands.sublime-commands:
```diff
@@ -20,15 +20,20 @@
 		"args": { "paths": [] }
 	},
+	{
+		"caption": "New…",
```
Final thing: this needs a `File: ` prefix, because out of context, in the command palette, it's not clear what it will do.

braver commented:
Pretty awesome! 💪🏻

mandx committed:
Add `File:` prefix to the entry in the Command Palette

mandx committed:
Few "pythonic" tweaks

- Do `len` checks first (since they are faster) in boolean contexts
- Replace `type(a) == A` with `isinstance(a, A)`

braver commented:
Let's ship it!
